### PR TITLE
Add integration tests for S3 storage driver

### PR DESCRIPTION
**Pull request by c-w**
### Description

This pull request adds integration tests for the S3 storage driver.

Given that there are existing buckets in the test AWS account, this pull request also updates the base tests to ignore any containers that weren't created as part of the test.

Unlike for Azure and MinIO, on S3, `create_container` for an already existing container seems to not throw `ContainerAlreadyExistsError` so the assertion has been disabled in this driver's test suite. In the future we may want to consider fixing all these minor inconsistencies that are brought to light by the tests.

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide)
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes)

### CODE REVIEW

File: integration/storage/base.py [hunk 1]
```diff
@@ -284,6 +292,16 @@ def _create_tempfile(self, prefix='', content=b''):
             self.addCleanup(os.remove, path)
             return path
 
+        @classmethod
+        def _random_container_name(cls):
+            suffix = random_string(cls.container_name_max_length)
+            name = cls.container_name_prefix + suffix
```
Comment (Kami) [inline integration/storage/base.py:298]: Yeah, I think including a special prefix on which we can filter on in the "clean up phase" is a good idea.

File: integration/storage/test_s3.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        for key, value in config.items():
+            if not value:
+                raise unittest.SkipTest('missing environment variable %s' % key)
+
+        cls.account = config['AWS_ACCESS_KEY_ID']
+        cls.secret = config['AWS_ACCESS_KEY_SECRET']
+
+    @classmethod
+    def tearDownClass(cls):
+        client = boto3.Session(
```
Comment (Kami) [inline integration/storage/test_s3.py:53]: Just curious why we use boto3 for the clean up phase and not the Libcloud itself?

**Review by Kami — APPROVED** [review]

**Comment by Kami** [comment]
I see the overall integration tests duration is ~14 minutes and ``test_objects_stream_iterable`` takes ~ 90 seconds.

Is there any way we could speed that up a bit? E.g. use a smaller object for that test or similar. Or we are intentionally testing large object size and we can't avoid slower duration?

**Comment by Kami** [comment]
> Unlike for Azure and MinIO, on S3, create_container for an already existing container seems to not throw ContainerAlreadyExistsError so the assertion has been disabled in this driver's test suite. In the future we may want to consider fixing all these minor inconsistencies that are brought to light by the tests.

Yeah I agree, this seems like something we should eventually fix for consistency across drivers. Please feel free to open an issue for it if you get a chance.

On that note, what is the current behavior? It just returns an existing container with that name?

File: integration/storage/test_s3.py (hunk 2 above)
Comment (c-w) [reply integration/storage/test_s3.py:53]: I figured that if we have a bug/regression in libcloud, we don't want to keep stray resources around so might as well use the official library to perform the cleanup.

**Comment by c-w** [comment]
> On that note, what is the current behavior? It just returns an existing container with that name?

Yes, `create_container` on an existing container returns a container with the same properties as the existing one. Here's a repro:

```py
from libcloud.storage.types import Provider
from libcloud.storage.providers import get_driver

cls = get_driver(Provider.S3)
driver = cls(AWS_ACCESS_KEY_ID, AWS_ACCESS_KEY_SECRET)

containers = driver.list_containers()
print(driver.create_container(containers[0].name).name == containers[0].name)
```

**Pull request merged**